src/PlaywrightSharp.Chromium/ChromiumNetworkManager.cs

```diff
@@ -23,5 +40,188 @@ internal Task SetUserAgentAsync(string userAgent)
         internal void InstrumentNetworkEvents(ChromiumSession session)
         {
         }
+
+        internal Task SetCacheEnabledAsync(bool enabled)
+        {
+            _userCacheDisabled = !enabled;
+            return UpdateProtocolCacheDisabledAsync();
+        }
+
+        private async void Client_MessageReceived(object sender, IChromiumEvent e)
+        {
+            try
+            {
+                switch (e)
+                {
+                    case NetworkResponseReceivedChromiumEvent networkResponseReceived:
+                        OnResponseReceived(networkResponseReceived);
+                        break;
+                    case NetworkRequestWillBeSentChromiumEvent networkRequestWillBeSent:
+                        OnRequestWillBeSent(networkRequestWillBeSent);
+                        break;
+                    case FetchRequestPausedChromiumEvent fetchRequestPaused:
+                        await OnRequestPausedAsync(fetchRequestPaused).ConfigureAwait(false);
+                        break;
+                    case NetworkLoadingFinishedChromiumEvent networkLoadingFinished:
+                        OnLoadingFinished(networkLoadingFinished);
+                        break;
+                }
+            }
+            catch (Exception ex)
+            {
+                // TODO Add Logger
+                /*
+                var message = $"Page failed to process {e.MessageID}. {ex.Message}. {ex.StackTrace}";
+                _logger.LogError(ex, message);
+                */
+                System.Diagnostics.Debug.WriteLine(ex.Message);
+                _client.OnClosed(ex.Message);
+            }
+        }
+
+        private void OnLoadingFinished(NetworkLoadingFinishedChromiumEvent e)
+        {
+            // For certain requestIds we never receive requestWillBeSent event.
+            // @see https://crbug.com/750469
+            if (!_requestIdToRequest.TryGetValue(e.RequestId, out var request))
+            {
+                return;
+            }
+
+            // Under certain conditions we never get the Network.responseReceived
+            // event from protocol. @see https://crbug.com/883475
+            var response = request.Request.Response;
+            if (response != null)
+            {
+                response.RequestFinished();
+            }
+
+            _requestIdToRequest.Remove(request.RequestId);
+            if (!string.IsNullOrEmpty(request.InterceptionId))
+            {
+                _attemptedAuthentications.Remove(request.InterceptionId);
+            }
+
+            _page.FrameManager.RequestFinished(request.Request);
+        }
+
+        private void OnRequest(NetworkRequestWillBeSentChromiumEvent e, string interceptionId)
+        {
+            if (e.Request.Url.StartsWith("data:"))
+            {
+                return;
+            }
+
+            var redirectChain = Array.Empty<Request>();
+            /*TODO

```

Comment:
todo?

Reply:
Scope. I will solve that when I solve Redirect tests.